### PR TITLE
Rewrite alignment newsletter module

### PR DESCRIPTION
**Pull request by Lovkush-A**
initially intention was to only introduce typehints, but I ended up making many changes so better to have separate PR.

note that there should be no changes to functionality with this pr

Changes include:
- adding docstrings and typehints and comments
- adding global constants for numerical constants
- renaming variables / methods
- simplifying (in my view) the sequence of method calls. instead of method1 calling method2 calling method3 calling method4, now have method1 which calls method2, method 3 and method 4 in a sequence.
- adding two test cases
- moving Item class outside of main class.
- reducing indentation levels for easier readability (in my opinion)

### CODE REVIEW

**Comment by Lovkush-A** [comment]
there is unusual bug with s, test yourself, where the test response is somehow lost.

```
TEST_RESPONSE 25: This seems relevant:
Adversarial Examples Are Not Bugs, They Are Features
...
TEST_RESPONSE 26: No matches found in the Alignment Newsletter Database
Science doesn't know
...
QUESTION # 25: FAILED
The sent message was 'TEST_QUESTION 25: paper search bugs features'
the expected message was 'REGULAR EXPRESSION: This seems relevant:\n.?Adversarial Examples Are Not Bugs, They Are Features'
the received message was 'NEVER RECEIVED A RESPONSE'
QUESTION # 26: FAILED
The sent message was 'TEST_QUESTION 26: blog post search: thisisnotarealword'
the expected message was 'No matches found in the Alignment Newsletter Database'
the received message was 'NEVER RECEIVED A RESPONSE'
```

File: modules/AlignmentNewsletterSearch.py [hunk 1]
```diff
@@ -5,114 +5,209 @@
...
+        )
+
+    def process_message(self, message: ServiceMessage) -> Response:
+        """Process a message and return a response if this module can handle it."""
+        text = self.is_at_me(message)
+        
+        if text is False:
+            return Response()
+
+        match = re.match(self.re_search, text)
```
Comment (MatthewBaggins) [inline modules/AlignmentNewsletterSearch.py:86]: you can just do
```py
match = self.re_search.match(text)
```

**Review by MatthewBaggins — APPROVED** [review]

File: modules/AlignmentNewsletterSearch.py [hunk 2]
```diff
@@ -5,114 +5,209 @@
...
+# this is the URL to Alignment Newsletter's google sheets database.
+SPREADSHEET_URL = "https://docs.google.com/spreadsheets/d/1PwWbWZ6FPqAgZWOoOcXM8N_tUCuxpEyMbN1NYYC02aM/export?format=zip"
+
+# weight factor to put on an item if it is a highlight
+HIGHLIGHT_WEIGHT = 1.5
+
+# maximum number of items to return in the response
+MAX_NUM_ITEMS = 5
+
+
```
Comment (MatthewBaggins) [inline modules/AlignmentNewsletterSearch.py:19]: I would turn `Item` into a `dataclass` and give it an alternative constructor `parse` that would be triggered in the loop in `load_items`

```py
@classmethod
def parse(cls, all_necessary_arguments) -> Optional[Item]:
    # do stuff to get all arguments needed for Item
    # if can't make item, return early with implicit None
    return Item(...)

# ...
# in AlignmentNewsletter module

def load_items(self) -> list[Item]:
    items: list[Item] = []
    for row in rows:
        item = Item.parse(all_relevant_args)
        if item is not None:
            items.append(item)
```

File: modules/AlignmentNewsletterSearch.py [hunk 3]
```diff
@@ -5,114 +5,209 @@
...
+        self.url = ""
+        self.title = ""
+        self.authors = ""
+        self.summary = ""
+        self.opinion = ""
+
+        self.score = 0
+
+    def __repr__(self):
+        return '<Item: %f "%s" %d>' % (self.score, self.title, len(self.summary))
```
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply modules/AlignmentNewsletterSearch.py:39]: agreed. now done

File: modules/AlignmentNewsletterSearch.py [hunk 4]
```diff
@@ -5,114 +5,209 @@
...
+
+    def __str__(self):
+        return self.__repr__()
```
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply modules/AlignmentNewsletterSearch.py:42]: quick google search does not give suggestion either way, so left as is

File: modules/AlignmentNewsletterSearch.py [hunk 5]
```diff
@@ -129,94 +224,115 @@ def extract_keywords(query):
...
-
-        return sorted(items, key=(lambda v: v.score), reverse=reverse)
-
-    def search(self, query):
-        result = self.sort_by_relevance(self.items, query, reverse=True)
-        self.log.info(self.class_name, search_results=result[:5])
-
-        best_score = result[0].score
+                item.score += (
+                    1.0 * item.title.lower().count(keyword) / (len(item.title) + 1)
```
[the remark this answers is not in the document]
Comment (Lovkush-A) [reply modules/AlignmentNewsletterSearch.py:254]: done

File: modules/AlignmentNewsletterSearch.py (hunk 1 above)
Comment (Lovkush-A) [reply modules/AlignmentNewsletterSearch.py:86]: actually just moved definition of regex pattern into this function, so now do re.match(pattern, text)

but did use your suggestion elsewhere (in parsing of rows with re_markdown_link)

File: modules/AlignmentNewsletterSearch.py (hunk 2 above)
Comment (Lovkush-A) [reply modules/AlignmentNewsletterSearch.py:19]: i was thinking of moving parsing into Item class, but wasn't sure how to make it work. I didn't think of using @classmethod! Thanks for suggestion. now implemented.

**Comment by Lovkush-A** [comment]
made most of aesthetic suggestions as suggested by @MatthewBaggins. thanks for the review!

**Pull request merged**